File: src/main/java/com/github/krystianmuchla/home/pagination/PaginatedResponse.java
package com.github.krystianmuchla.home.pagination;

import java.util.List;
import java.util.function.Function;

public record PaginatedResponse<T>(List<T> data, PaginationResponse pagination) {
    public <U> PaginatedResponse(final PaginatedResult<U> paginatedResult, final Function<U, T> mapper) {
        this(
            paginatedResult.data().stream().map(mapper).toList(),
            new PaginationResponse(paginatedResult.pagination())
        );
    }
}

File: src/main/kotlin/com/example/skyr/pagination/api/PaginatedResponse.kt
package com.example.skyr.pagination.api

import com.example.skyr.pagination.PaginatedResult

data class PaginatedResponse<T>(val data: List<T>, val pagination: PaginationResponse)

fun <T, U> paginatedResponse(paginatedResult: PaginatedResult<U>, transformer: (U) -> T) =
    PaginatedResponse(
        paginatedResult.data.stream().map { element -> transformer(element) }.toList(),
        paginationResponse(paginatedResult.pagination)
    )
